Improve counting of local solutions for QuadraticForm at p=2

WvanWoerden opened the pull request:
Fixes #38679.

Introduces new method `CountAllLocalGoodTypesNormalForm(Q, p, k, m, zvec, nzvec)` to compute the number of local solutions of 'good type' for Q[v] = m mod p^k when Q is in local normal form at p.
This replaces the use of the slow brute-force function `CountAllLocalTypesNaive` indirectly used in `local_good_density_congruence_even` for (p,k)=(2,3) . 

### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->

- [x] The title is concise and informative.
- [ ] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation and checked the documentation preview.

### :hourglass: Dependencies

WvanWoerden committed:
Add CountAllLocalGoodTypesNormalForm, use in local_good_density_congruence_even

fchapoton commented on src/sage/quadratic_forms/count_local_2.pyx:
```diff
@@ -228,27 +228,27 @@ cdef local_solution_type_cdef(Q, p, w, zvec, nzvec):
 
     if not nonzero_flag:
-        return <long> 0
+        return < long > 0
```
this is not wanted, keep `<long>`

fchapoton commented on src/sage/quadratic_forms/count_local_2.pyx:
```diff
@@ -271,13 +271,93 @@ cdef local_solution_type_cdef(Q, p, w, zvec, nzvec):
     raise RuntimeError("Error in IsLocalSolutionType: Should not execute this line... =( \n")
+
+
+def CountAllLocalGoodTypesNormalForm(Q, p, k, m, zvec, nzvec):
```
do not use CamelCase name for functions

WvanWoerden committed:
Fix doctest

The old doctest failed the input assumption that Q is given in local diagonal form. Now gives such a local diagonal form to the function. The new output has also been verified with the old code.

WvanWoerden committed:
Revert wrong autoformatting of cython code

WvanWoerden committed:
Rename function

fchapoton committed:
Merge branch 'develop' into 2_local_counting

WvanWoerden committed:
Merge branch 'develop' into 2_local_counting

S17A05 commented on src/sage/quadratic_forms/count_local_2.pyx:
```diff
@@ -281,3 +281,83 @@ cdef local_solution_type_cdef(Q, p, w, zvec, nzvec):
+
+    OUTPUT:
+
+    an integer '\ge 0' representing the solutions of Good type.
```
A few things here:
- If I see it correctly, m_range is always p^k here, so it might be nicer to write "mod `p^k`" in the first line
- In the output line '\ge 0' is not displaying properly since the wrong ticks are used; it should be `\ge 0`, or alternatively you could also write "a non-negative integer" instead (I think the latter is slightly better, but either should be fine)
- Not sure if it is common to say "representing the solutions" for the count of solutions, but maybe "giving the number of solutions" would be a bit more explicit?

WvanWoerden committed:
Fix docs
